### PR TITLE
Add an exponentiation trait

### LINKED ISSUE #1273: Add and implement Exponentiation Trait
We should provide Sway users convenient methods for calculating exponents of values.

This would take the form of a wrapper around the `exp` instruction, most likely in `core::ops`.

Also a slack comment here re a dedicated token for this op: https://fuellabs.slack.com/archives/C01KBNGSWGJ/p1650642887038259

### PR DESCRIPTION
**Pull request by nfurfaro**
closes #1273

### CODE REVIEW

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (adlerjohn) [inline test/src/sdk-harness/test_projects/exponentiation/mod.rs:30]: Remove?

Comment (nfurfaro) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:30]: yes

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (vnepveu) [inline test/src/sdk-harness/test_projects/exponentiation/mod.rs:1]: wouldn't this file be better named `harness.rs`? `mod` seems weird to me.

Comment (vnepveu) [inline test/src/sdk-harness/test_projects/exponentiation/mod.rs:1]: Remove this dependency and the associated test for now, it's not ready in the SDK.

**Review by vnepveu — CHANGES_REQUESTED** [review]
`script_runner` is not ready, otherwise LGTM!

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (adlerjohn) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:1]: It's `mod.rs` for all the SDK harness tests, so orthogonal to this PR.

Comment (nfurfaro) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:1]: This is how we do it atm in the stdlib test suite.
then, in `harness.rs` we collect all the mods like
```
// Add test modules here:

mod auth;
mod call_frames;
mod context;
mod exponentiation;
mod hashing;
mod logging;
mod reentrancy;
mod registers;
mod storage;
mod token_ops;
mod tx_fields;

```

Comment (nfurfaro) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:1]: This test helper is local the stdlib. It has been deprecated though.

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (adlerjohn) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs]: Fixed in [e8ba47e](https://github.com/FuelLabs/sway/pull/1744/commits/e8ba47e5ff056a586ee2a6fb9f0c7979f87de88c)

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (vnepveu) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:1]: oops got caught up in the SDK haha

Comment (vnepveu) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:1]: got it thanks

Comment (adlerjohn) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs]: Fixed in [712391d](https://github.com/FuelLabs/sway/pull/1744/commits/712391df072afbe63e50765559aae2f1519df015)

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (nfurfaro) [inline test/src/sdk-harness/test_projects/exponentiation/mod.rs:46]: This can be removed.

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (nfurfaro) [inline test/src/sdk-harness/test_projects/exponentiation/mod.rs:47]: now this should panic

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (nfurfaro) [inline test/src/sdk-harness/test_projects/exponentiation/mod.rs:69]: Same as above

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (adlerjohn) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:47]: What do you mean?

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (adlerjohn) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:46]: Keeping these tests around will allow us to catch regressions in expected behavior from the VM.

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (adlerjohn) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:47]: 10^5 fits withing a `u64`, so it would not panic until the mentioned issue lands.

Comment (nfurfaro) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:47]: This test is expected to panic, but the `should_panic` is commented out because of a former issue.

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (nfurfaro) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:46]: I just mean the TODO can be removed.

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (adlerjohn) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:47]: Are you saying that the `#[should_panic(expected = "ArithmeticOverflow")]` line should be uncommented?

Comment (nfurfaro) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:47]: ahh. k nvm :)

File: test/src/sdk-harness/test_projects/exponentiation/mod.rs
Comment (nfurfaro) [reply test/src/sdk-harness/test_projects/exponentiation/mod.rs:46]: nvm.

**Comment by nfurfaro** [comment]
I would approve when CI completes, but I can't approve as I'm the author.

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/exponentiation_test/src/main.sw
Comment (mohammadfawaz) [inline test/src/e2e_vm_tests/test_programs/should_pass/stdlib/exponentiation_test/src/main.sw:3]: This file just needs formatting. Looks good otherwise!

**Review by vnepveu — DISMISSED** [review]
Lgtm

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/exponentiation_test/src/main.sw
Comment (adlerjohn) [reply test/src/e2e_vm_tests/test_programs/should_pass/stdlib/exponentiation_test/src/main.sw]: Fixed in [e1f8702](https://github.com/FuelLabs/sway/pull/1744/commits/e1f8702185aeedf3300f4b3cb107eb20acc6cedc)

**Review by mohammadfawaz — APPROVED** [review]

**Pull request merged**
